java/src/jmri/jmrit/decoderdefn/DecoderIndexFile.java

```diff
@@ -443,11 +446,38 @@ public static void forceCreationOfNewIndex(boolean increment) {
         }
 
         // write it out
-        try {
-            index.writeFile(DECODER_INDEX_FILE_NAME, InstanceManager.getDefault(DecoderIndexFile.class), sbox);
-        } catch (java.io.IOException ex) {
-            log.error("Error writing new decoder index file: {}", ex.getMessage());
-        }
+        if (sbox.length < 30 || GraphicsEnvironment.isHeadless()) {
+            try {
+                index.writeFile(DECODER_INDEX_FILE_NAME,
+                            InstanceManager.getDefault(DecoderIndexFile.class), sbox, null);
+            } catch (java.io.IOException ex) {
+                log.error("Error writing new decoder index file: {}", ex.getMessage());
+            }
+            return;
```

Comment:
At line 840, there's an `updateIndexIfNeeded()` call that can invoke this `forceCreationOfNewIndex`. When it returns, a new instance of the `DecoderIndexFile` is created and put into use.  This happens when you have a local decoder index that's out of date with respect to the one in the program directory, typically because a new JMRI version has been installed.  When this code takes the process-and-immediate return path, that works.  But when the separate thread is used, the new instance will be created before the new index file has actually been created, so the new one won't get used until there's a restart.

Let me think a little about how we can handle this.  Might be sufficient to turn the automated check into a warning message, or maybe add the instance update on to the end of the thread.

Reply:
Hmmm, yeah. It looks like that might happen in a few places. I'd say we could add a parameter to specify to load the file while blocking, but that means no user visible interaction, and I'm not sure how we'd tell one case from the next.
Maybe doing the instance update at the end of the thread would be best (similar to the roster index). And then the progress dialog could have "reloading new decoder index" while that's happening.

I can look at what it would take to add the load in the thread, unless you're already figuring it out... :)

Reply:
To see this in practice, (1) recreate decoder index to make sure there's a decoderIndex.xml file in the profile directory (2) Edit that decoderIndex.xml file to lower the version attribute on the decoderIndex element, e.g.
```
  <decoderIndex version="70">
```
When you restart the program, you'll see the progress bar as it automatically recreates the index.

To fix the problem with the re-index's result not being used, I think all you have to do is to add
```
try {
    var instance = new DecoderIndexFile();
    instance.readFile(defaultDecoderIndexFilename());
    InstanceManager.store(instance, DecoderIndexFile.class);
} catch (IOException | JDOMException e) {
    log.error("Exception during decoder index reload: ", e);
}
```
at the end of the thread's execution, e.g. when you're closing out `pm`.

That will reload both the local `instance` variable and also the default value that's available through the InstanceManager.

Reply:
@bobjacobsen 
How about showing a modal dialog when this happens? The modal dialog will prevent the user from doing anything with JMRI while the decoder index is created, but the modal dialog will be able to show the progress bar. Once the decoder index is created, the decoder index would be reloaded.

Reply:
@danielb987 the problem is that any UI components won't be updated until the method returns, because it's running in the event dispatch thread. So we need to create a background thread that updates the dialog as it progresses, so the event dispatch thread can actually display and update the progress dialog.

The hard part here is there are various calls to get the decoder index that might call this method, and with a background thread the method will return before the new index is created and ready. We may need a singleton boolean or somesuch that says "index is being recreated" which the parent methods check, and if it's true, wait until it's false before continuing. But that in itself is problematic because that's in the event dispatch thread, so same problem (no UI updates).
There may be a way to do it using SwingUtils.invokeLater() but I haven't quite worked out that logic.

Reply:
@bobjacobsen as I said above, even if we reload the decoder index at the end of the thread, the calling logic in the event dispatch thread may have already read the old decoder index and acted upon it. Hmmm.

Reply:
@danielb987 on second thought you may be right about the modal dialog. I'll look into it. It's just more coding for something that _should_ be simple :)

Reply:
I think I'm going to close this PR and modify the logic to use SwingWorker with a modal dialog that will block the execution until the worker completes, similar to the code mentioned in get():
https://docs.oracle.com/javase/7/docs/api/javax/swing/SwingWorker.html#get()